File: java/src/test/java/md/remit/ComplianceTest.java
package md.remit;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import md.remit.models.Transaction;
import md.remit.models.TransactionList;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.concurrent.atomic.AtomicReference;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assumptions.assumeTrue;

/**
 * Compliance tests: Java SDK against a real running server.
 *
 * <p>Tests are skipped when the server is unreachable. Boot the server with:
 * <pre>
 *   docker compose -f docker-compose.compliance.yml up -d
 * </pre>
 *
 * <p>Environment variables:
 * <ul>
 *   <li>REMIT_TEST_SERVER_URL (default: http://localhost:3000)</li>
 *   <li>REMIT_ROUTER_ADDRESS  (default: 0x70997970C51812dc3A010C7d01b50e0d17dc79C8)</li>
 * </ul>
 */
class ComplianceTest {

    private static final String SERVER_URL =
        envOrDefault("REMIT_TEST_SERVER_URL", "http://localhost:3000");
    private static final String ROUTER_ADDRESS =
        envOrDefault("REMIT_ROUTER_ADDRESS", "0x70997970C51812dc3A010C7d01b50e0d17dc79C8");

    private static final ObjectMapper MAPPER = new ObjectMapper();
    private static final HttpClient HTTP = HttpClient.newBuilder()
        .connectTimeout(Duration.ofSeconds(5))
        .build();

    /** Set to false when the server health-check fails - all tests skip. */
    private static boolean serverAvailable = false;

    /** Shared funded payer wallet - only one mint per test run. */
    private static final AtomicReference<Wallet> SHARED_PAYER = new AtomicReference<>();
    private static final AtomicReference<String> SHARED_PAYER_ADDR = new AtomicReference<>();

    @BeforeAll
    static void checkServerAvailability() {
        System.out.println("[COMPLIANCE] checking server availability at " + SERVER_URL + "/health");
        try {
            HttpRequest req = HttpRequest.newBuilder()
                .uri(URI.create(SERVER_URL + "/health"))
                .timeout(Duration.ofSeconds(3))
                .GET()
                .build();
            HttpResponse<Void> resp = HTTP.send(req, HttpResponse.BodyHandlers.discarding());
            serverAvailable = resp.statusCode() == 200;
            System.out.println("[COMPLIANCE] server health check: status=" + resp.statusCode() + " available=" + serverAvailable);
        } catch (Exception ignored) {
            serverAvailable = false;
            System.out.println("[COMPLIANCE] server health check: UNREACHABLE (tests will be skipped)");
        }
    }

    // ─── Helpers ──────────────────────────────────────────────────────────────

    private static String envOrDefault(String key, String defaultValue) {
        String val = System.getenv(key);
        return (val != null && !val.isBlank()) ? val : defaultValue;
    }

    private static String post(String path, String jsonBody, String bearerToken) throws Exception {
        HttpRequest.Builder builder = HttpRequest.newBuilder()
            .uri(URI.create(SERVER_URL + path))
            .timeout(Duration.ofSeconds(10))
            .header("Content-Type", "application/json")
            .POST(HttpRequest.BodyPublishers.ofString(jsonBody));
        if (bearerToken != null) {
            builder.header("Authorization", "Bearer " + bearerToken);
        }
        HttpResponse<String> resp = HTTP.send(builder.build(), HttpResponse.BodyHandlers.ofString());
        return resp.body();
    }

    private static String getWithAuth(String path, String bearerToken) throws Exception {
        HttpResponse<String> resp = HTTP.send(
            HttpRequest.newBuilder()
                .uri(URI.create(SERVER_URL + path))
                .timeout(Duration.ofSeconds(10))
                .header("Authorization", "Bearer " + bearerToken)
                .GET()
                .build(),
            HttpResponse.BodyHandlers.ofString()
        );
        return resp.body();
    }

    /** Generate a random private key and derive the wallet address. */
    private static String[] registerAndGetKey() throws Exception {
        System.out.println("[COMPLIANCE] generating random private key...");
        byte[] keyBytes = new byte[32];
        new java.security.SecureRandom().nextBytes(keyBytes);
        StringBuilder sb = new StringBuilder("0x");
        for (byte b : keyBytes) {
            sb.append(String.format("%02x", b));
        }
        String privateKey = sb.toString();

        Wallet wallet = RemitMd.withKey(privateKey)
            .chain("base")
            .testnet(true)
            .baseUrl(SERVER_URL)
            .routerAddress(ROUTER_ADDRESS)
            .build();

        System.out.println("[COMPLIANCE] wallet created: " + wallet.address() + " (chain=" + wallet.chainId() + ")");
        return new String[]{privateKey, wallet.address()};
    }

    /** Fund a wallet via mint (no auth required on testnet). */
    private static void fundWallet(String walletAddress) throws Exception {
        System.out.println("[COMPLIANCE] mint: 1000 USDC -> " + walletAddress);
        String body = MAPPER.writeValueAsString(
            java.util.Map.of("wallet", walletAddress, "amount", 1000));
        JsonNode resp = MAPPER.readTree(post("/api/v1/mint", body, null));
        assertThat(resp.has("tx_hash"))
            .as("mint response must contain tx_hash, got: " + resp)
            .isTrue();
        System.out.println("[COMPLIANCE] mint: 1000 USDC -> " + walletAddress + " tx=" + resp.get("tx_hash").asText());
    }

    /** Return (or lazily create) the shared funded payer wallet. */
    private static Wallet getSharedPayer() throws Exception {
        if (SHARED_PAYER.get() != null) {
            System.out.println("[COMPLIANCE] reusing shared payer: " + SHARED_PAYER.get().address());
            return SHARED_PAYER.get();
        }
        System.out.println("[COMPLIANCE] creating shared payer wallet...");
        String[] keyAddr = registerAndGetKey();
        String privateKey = keyAddr[0];
        String walletAddr = keyAddr[1];
        fundWallet(walletAddr);
        Wallet payer = RemitMd.withKey(privateKey)
            .chain("base")
            .testnet(true)
            .baseUrl(SERVER_URL)
            .routerAddress(ROUTER_ADDRESS)
            .build();
        SHARED_PAYER.set(payer);
        SHARED_PAYER_ADDR.set(walletAddr);
        System.out.println("[COMPLIANCE] shared payer ready: " + payer.address() + " (funded 1000 USDC)");
        return payer;
    }

    // ─── Auth tests ───────────────────────────────────────────────────────────

    @Test
    void authenticatedRequest_returnsWalletBalance_not401() throws Exception {
        assumeTrue(serverAvailable, "Compliance server not reachable at " + SERVER_URL);
        System.out.println("[COMPLIANCE] test: authenticatedRequest_returnsWalletBalance_not401");

        String[] keyAddr = registerAndGetKey();
        Wallet wallet = RemitMd.withKey(keyAddr[0])
            .chain("base")
            .testnet(true)
            .baseUrl(SERVER_URL)
            .routerAddress(ROUTER_ADDRESS)
            .build();
        System.out.println("[COMPLIANCE] auth wallet created: " + wallet.address() + " (chain=" + wallet.chainId() + ")");

        // reputation() makes an authenticated GET to /api/v1/reputation/{address} -
        // this endpoint exists for all registered addresses and returns 401 if auth fails.
        System.out.println("[COMPLIANCE] calling reputation(" + wallet.address() + ") with auth...");
        var rep = wallet.reputation(wallet.address());
        assertThat(rep).as("reputation() must not throw 401").isNotNull();
        System.out.println("[COMPLIANCE] reputation returned successfully (auth verified, no 401)");
    }

    @Test
    void unauthenticatedRequest_returns401() throws Exception {
        assumeTrue(serverAvailable, "Compliance server not reachable at " + SERVER_URL);
        System.out.println("[COMPLIANCE] test: unauthenticatedRequest_returns401");

        System.out.println("[COMPLIANCE] sending unauthenticated POST /api/v1/payments/direct...");
        HttpRequest req = HttpRequest.newBuilder()
            .uri(URI.create(SERVER_URL + "/api/v1/payments/direct"))
            .timeout(Duration.ofSeconds(5))
            .header("Content-Type", "application/json")
            .POST(HttpRequest.BodyPublishers.ofString(
                "{\"to\":\"0x70997970C51812dc3A010C7d01b50e0d17dc79C8\",\"amount\":\"1.000000\"}"))
            .build();
        HttpResponse<String> resp = HTTP.send(req, HttpResponse.BodyHandlers.ofString());
        System.out.println("[COMPLIANCE] unauthenticated response: status=" + resp.statusCode() + " body=" + resp.body());
        assertThat(resp.statusCode())
            .as("unauthenticated POST /payments/direct must return 401")
            .isEqualTo(401);
        System.out.println("[COMPLIANCE] 401 confirmed for unauthenticated request");
    }

    // ─── Payment tests ────────────────────────────────────────────────────────

    @Test
    void payDirect_happyPath_returnsTxHash() throws Exception {
        assumeTrue(serverAvailable, "Compliance server not reachable at " + SERVER_URL);
        System.out.println("[COMPLIANCE] test: payDirect_happyPath_returnsTxHash");

        Wallet payer = getSharedPayer();
        String[] payeeKeyAddr = registerAndGetKey();
        String payeeAddr = payeeKeyAddr[1];
        System.out.println("[COMPLIANCE] payee wallet created: " + payeeAddr);

        try {
            System.out.println("[COMPLIANCE] pay: 5.0 USDC " + payer.address() + " -> " + payeeAddr + " memo=\"java compliance test\"");
            Transaction tx = payer.pay(payeeAddr, new BigDecimal("5.0"), "java compliance test");
            assertThat(tx.txHash)
                .as("pay() must return a non-empty tx_hash")
                .isNotEmpty();
            System.out.println("[COMPLIANCE] pay: 5.0 USDC " + payer.address() + " -> " + payeeAddr
                + " tx=" + tx.txHash + " id=" + tx.id + " fee=" + tx.fee);
        } catch (RemitError e) {
            System.err.println("[COMPLIANCE] pay() FAILED: code=" + e.getCode()
                + " message=" + e.getMessage() + " httpStatus=" + e.getHttpStatus()
                + " payerAddr=" + payer.address());
            throw e;
        }
    }

    @Test
    void payDirect_belowMinimum_throwsRemitError() throws Exception {
        assumeTrue(serverAvailable, "Compliance server not reachable at " + SERVER_URL);
        System.out.println("[COMPLIANCE] test: payDirect_belowMinimum_throwsRemitError");

        Wallet payer = getSharedPayer();
        String[] payeeKeyAddr = registerAndGetKey();
        String payeeAddr = payeeKeyAddr[1];
        System.out.println("[COMPLIANCE] payee wallet created: " + payeeAddr);

        RemitError caught = null;
        try {
            System.out.println("[COMPLIANCE] pay: 0.0001 USDC " + payer.address() + " -> " + payeeAddr + " (expect rejection)");
            payer.pay(payeeAddr, new BigDecimal("0.0001"), "too small");
        } catch (RemitError e) {
            caught = e;
            System.out.println("[COMPLIANCE] below-minimum correctly rejected: code=" + e.getCode()
                + " message=" + e.getMessage() + " httpStatus=" + e.getHttpStatus());
        }
        assertThat(caught)
            .as("pay() with amount below minimum must throw RemitError")
            .isNotNull();
        System.out.println("[COMPLIANCE] below-minimum test passed");
    }
}

File: elixir/test/compliance_test.exs
defmodule RemitMd.ComplianceTest do
  # Compliance tests: Elixir SDK against a real running server.
  #
  # Tests are tagged @tag :compliance and skipped when the server is unreachable.
  # Boot the server with:
  #   docker compose -f docker-compose.compliance.yml up -d
  #
  # Environment variables:
  #   REMIT_TEST_SERVER_URL  (default: http://localhost:3000)
  #   REMIT_ROUTER_ADDRESS   (default: 0x70997970C51812dc3A010C7d01b50e0d17dc79C8)
  use ExUnit.Case, async: false

  alias RemitMd.Wallet

  @server_url System.get_env("REMIT_TEST_SERVER_URL", "http://localhost:3000")
  @router_address System.get_env(
    "REMIT_ROUTER_ADDRESS",
    "0x70997970C51812dc3A010C7d01b50e0d17dc79C8"
  )

  # ─── Shared state ─────────────────────────────────────────────────────────

  # Agent holds the shared funded payer key (pk) so we only call mint once.
  defmodule SharedPayer do
    use Agent

    def start_link(_opts), do: Agent.start_link(fn -> nil end, name: __MODULE__)
    def get, do: Agent.get(__MODULE__, & &1)
    def set(val), do: Agent.update(__MODULE__, fn _ -> val end)
  end

  setup_all do
    {:ok, _} = SharedPayer.start_link([])
    :ok
  end

  # ─── Helpers ──────────────────────────────────────────────────────────────

  defp server_available? do
    :inets.start()
    :ssl.start()
    url = ~c"#{@server_url}/health"

    case :httpc.request(:get, {url, []}, [{:timeout, 3000}], []) do
      {:ok, {{_, 200, _}, _, _}} -> true
      _ -> false
    end
  end

  defp http_post(path, body, token \\ nil) do
    :inets.start()
    :ssl.start()
    url = ~c"#{@server_url}#{path}"
    json = Jason.encode!(body)
    headers = [
      {~c"content-type", ~c"application/json"}
      | if(token, do: [{~c"authorization", ~c"Bearer #{token}"}], else: [])
    ]

    {:ok, {{_, status, _}, _headers, resp_body}} =
      :httpc.request(
        :post,
        {url, headers, ~c"application/json", json},
        [{:timeout, 10_000}],
        []
      )

    {status, Jason.decode!(to_string(resp_body))}
  end

  defp http_get(path, token) do
    :inets.start()
    :ssl.start()
    url = ~c"#{@server_url}#{path}"
    headers = [{~c"authorization", ~c"Bearer #{token}"}]

    {:ok, {{_, status, _}, _headers, resp_body}} =
      :httpc.request(:get, {url, headers}, [{:timeout, 10_000}], [])

    {status, Jason.decode!(to_string(resp_body))}
  end

  defp register_and_get_key do
    ts = :os.system_time(:millisecond)
    email = "compliance.elixir.#{ts}@test.remitmd.local"

    {201, reg} =
      http_post("/api/v0/auth/register", %{email: email, password: "ComplianceTestPass1!"})

    token       = reg["token"] || raise "register failed: #{inspect(reg)}"
    wallet_addr = reg["wallet_address"] || raise "no wallet_address in register response"

    {200, key_data} = http_get("/api/v0/auth/agent-key", token)
    private_key = key_data["private_key"] || raise "agent-key failed: #{inspect(key_data)}"

    {private_key, wallet_addr}
  end

  defp fund_wallet(wallet_addr) do
    {200, resp} = http_post("/api/v0/mint", %{wallet: wallet_addr, amount: 1000})
    assert resp["tx_hash"] != nil, "mint response must contain tx_hash, got: #{inspect(resp)}"
  end

  defp make_wallet(private_key) do
    Wallet.new(
      private_key: private_key,
      chain: "base_sepolia",
      api_url: @server_url <> "/api/v0",
      router_address: @router_address
    )
  end

  defp get_shared_payer do
    case SharedPayer.get() do
      nil ->
        {pk, addr} = register_and_get_key()
        fund_wallet(addr)
        SharedPayer.set(pk)
        make_wallet(pk)

      pk ->
        make_wallet(pk)
    end
  end

  # ─── Auth tests ───────────────────────────────────────────────────────────

  test "authenticated request returns balance, not 401" do
    if not server_available?() do
      IO.puts("SKIP: compliance server not reachable at #{@server_url}")
      :ok
    else
      wallet = get_shared_payer()

      # reputation/2 makes an authenticated GET to /api/v0/reputation/{address} —
      # this endpoint exists for all registered addresses and returns 401 if auth fails.
      {:ok, rep} = Wallet.reputation(wallet, wallet.address)
      assert rep != nil
    end
  end

  test "unauthenticated POST /payments/direct returns 401" do
    if not server_available?() do
      :ok
    else
      {status, _body} =
        http_post("/api/v0/payments/direct", %{
          to: "0x70997970C51812dc3A010C7d01b50e0d17dc79C8",
          amount: "1.000000"
        })

      assert status == 401, "unauthenticated POST must return 401, got: #{status}"
    end
  end

  # ─── Payment tests ────────────────────────────────────────────────────────

  test "pay_direct happy path returns tx_hash" do
    if not server_available?() do
      :ok
    else
      payer = get_shared_payer()
      {_pk, payee_addr} = register_and_get_key()

      {:ok, tx} = Wallet.pay(payer, payee_addr, "5.0", description: "elixir compliance test")
      assert tx.tx_hash != nil
      assert tx.tx_hash != ""
    end
  end

  test "pay_direct below minimum returns error" do
    if not server_available?() do
      :ok
    else
      payer = get_shared_payer()
      {_pk, payee_addr} = register_and_get_key()

      result = Wallet.pay(payer, payee_addr, "0.0001")
      assert match?({:error, _}, result),
             "pay with amount below minimum must return {:error, ...}, got: #{inspect(result)}"
    end
  end
end
